docs/developer/architecture/development-plugin-saved-objects.asciidoc

```diff
@@ -202,4 +202,226 @@ up to date. If a visualization `id` was directly stored in
+        const chartMigrations = this.charts.reduce((migrations, chart) => {
+          for (const [version, chartMigration] of Object.entries(chart.getMigrations)) {
+            const existingMigration = migrations[version] ?? (input: any) => input;
+            migrations[version] = (chartDoc) => {
+              if (chartDoc.type === chart.type) {
+                chartDoc = chartMigration(chartDoc);
+              }
+
+              return existingMigration(chartDoc);
+            }
+          }
+        }, {} as Record<string, ChartMigrationFn>);
+        
+        return {
+          '1.1.0': (oldChart) => {
+            return {
+              ...oldChart,
+              // Apply the migrations on the chart state if and only if any charts define migrations for this version.
+              // It's important that the registry items can only access the `state` part of the complete chart object.
+              // This is the part of the document that the chart implementation functionally "owns".
+              state: chartMigrations['1.1.0'] ? chartMigrations['1.1.0'](oldChart.chartState) : oldChart.chartState
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+interface BarChartState100 {
+  xAxis: string;
+  yAxis: string;
+  indexPattern: string;
+}
+
+interface BarChartState110 {
+  xAxis: string;
+  yAxis: string;
+  dataSource: {
+    indexPattern: string;
+  }
+}
+
+class BarChartPlugin {
+  public setup(core, plugins) {
+    plugins.charts.register<BarChartState100>({
+      type: 'bar-chart',
+      render() { ... },
+      create() { return { xAxis: 'foo', yAxis: 'bar', dataSource: { indexPattern: 'baz' } }},
+      getMigrations() {
+        return {
+          '1.1.0': (oldChartState: BarChartState100): BarChartState110 => ({
+            xAxis: oldChartState.xAxis,
+            yAxis: oldChartState.yAxis,
+            dataSource: { indexPattern: oldChartState.dataSource }
+          })
+        }
+      }
+    })
+  }
+}
+
+class ChartListPlugin {
+  public setup(core, plugins) {
+    core.savedObjects.registerType({
+      name: 'chart-list',
+      hidden: false,
+      mappings: {
+        dynamic: false,
+        properties: {
+          title: {
+            type: 'text',
+          },
+          // We will store charts as an array, no need to index this field.
+          charts: {
+            index: false;
+          }
+        },
+      },
+      getMigrations() {
+        // Request the migrations for the chart state from the charts registry
+        const chartMigrations = plugins.charts.getMigrations();
+
+        return {
+          '1.1.0': (chartList) => {
+            return {
+              ...chartList,
+              // For each chart, apply the chart migration for this version if it exists
+              charts: chartList.map(
+                chart => chartMigrations['1.1.0'] ? chartMigrations['1.1.0'](chart) : chart
+              )
+            }
```

Comment:
Is there a way we can automate this so the ChartList plugin author doesn't have to manually add a new migration in here every single minor? It'd be super easy to forget.

something like
```ts
 const migrationMap = {
    '1.1.0': my1_0_migration(),
     '2.5.0': my_2_5_migration()
  }

 // For versions that the Chart List plugin doesn't have any specific migrations for, they still want to
 // let each chart have an opportunity to add migrations.   Rather than require this to be manually added
 // every time, automatically fill in any versions that don't have manual migrations. 
  KIBANA_VERSION_LIST.foreach(version => {
    if (migrationMap[version] === undefined) {
      migrationMap[version] = {
              ...chartList,
              // For each chart, apply the chart migration for this version if it exists
              charts: chartList.map(
                chart => chartMigrations[version] ? chartMigrations[version](chart) : chart
             )
      }
    }
  );

  return migrationMap;
```

You would still need to remember to migrate the nested state inside each `my_x_migration` fn. Couple perhaps be remedied by some more refactoring.

Reply:
Or even make it a part of registry API:
```js
charts: chartList.map(state => plugins.charts.migrateTo(version, state))
```